src/dbnode/storage/shard.go

```diff
@@ -2660,74 +2661,216 @@ func (s *dbShard) Repair(
-	var lastWriteError error
-	var processedBlockCount int64
+	var (
+		processedTileCount atomic.Int64
+		multiErr           xerrors.MultiError
+		dp                 ts.Datapoint
+		segmentCapacity    int
+		writerData         = make([][]byte, 2)
+	)
 
-	for {
-		id, tags, data, _, err := reader.Read()
-		if err == io.EOF {
-			break
+	READER:
+	for readerIter.Next() {
+		seriesIter, id, encodedTags := readerIter.Current()
+
+		for seriesIter.Next() {
+			frame := seriesIter.Current()
+			unit, singleUnit := frame.Units().SingleValue()
+			annotation, singleAnnotation := frame.Annotations().SingleValue()
+
+			if frameValues := frame.Values(); len(frameValues) > 0 {
+				lastIdx := len(frameValues) - 1
+
+				value := frameValues[lastIdx]
+				timestamp := frame.Timestamps()[lastIdx]
+
+				if !singleUnit {
+					// TODO: what happens if unit has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Units().Values()) {
+						unit = frame.Units().Values()[lastIdx]
+					}
+				}
+
+				if !singleAnnotation {
+					// TODO: what happens if annotation has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Annotations().Values()) {
+						annotation = frame.Annotations().Values()[lastIdx]
+					}
+				}
+
+				dp.Timestamp = timestamp
+				dp.Value = value
+				err = encoder.Encode(dp, unit, annotation)
+				if err != nil {
+					s.metrics.largeTilesWriteErrors.Inc(1)
+					multiErr = multiErr.Add(err)
+					encoder.Reset(opts.Start, segmentCapacity, targetSchemaDescr)
+					continue READER
+				}
+
+				processedTileCount.Inc()
+			}
 		}
-		if err != nil {
-			return processedBlockCount, err
+
+		segment := encoder.DiscardReset(opts.Start, segmentCapacity, targetSchemaDescr)
+
+		segmentLen := segment.Len()
+		if segmentLen > segmentCapacity {
+			// Will use the same capacity for the next series.
+			segmentCapacity = segmentLen
 		}
 
-		data.IncRef()
-		bytesReader.Reset(data.Bytes())
-		dataPointIter.Reset(bytesReader, nil)
+		writerData[0] = segment.Head.Bytes()
+		writerData[1] = segment.Tail.Bytes()
+		checksum := segment.CalculateChecksum()
 
-		for dataPointIter.Next() {
-			dp, unit, annot := dataPointIter.Current()
-			_, err = s.writeAndIndex(ctx, id, tags, dp.Timestamp, dp.Value, unit, annot, wOpts, true)
-			if err != nil {
-				s.metrics.largeTilesWriteErrors.Inc(1)
-				lastWriteError = err
-			}
+		if err := writer.WriteAll(id.Bytes(), encodedTags, writerData, checksum); err != nil {
+			s.metrics.largeTilesWriteErrors.Inc(1)
+			multiErr = multiErr.Add(err)
+		} else {
+			s.metrics.largeTilesWrites.Inc(1)
 		}
 
-		dataPointIter.Close()
+		id.Finalize()
+		segment.Finalize()
+	}
+
+	if err := readerIter.Err(); err != nil {
+		multiErr = multiErr.Add(err)
+		if err := writer.Abort(); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	} else if err := writer.Close(); err != nil {
```

Comment:
Should we also not try to avoid writing a clean file if `multiErr.NumErrors() > 0`? Seems strange to write out an incomplete file (with missing series or if failed to write series), probably better to abort in that case too?

Reply:
Up to this point, the only way to accumulate the `multiErr` is from the `encoder` errors, and in case of `encoder` errors we do not abort the aggregation but continue to the next series instead. Which I believe contradicts your proposed logics of avoiding any partial results. 
So I guess in case of `encoder` error I should stop the aggregation entirely and call the `writer.Abort`?